yo1k/tic_tac_toe/test/core/test_game.py

```diff
@@ -0,0 +1,267 @@
+from yo1k.tic_tac_toe.core.game import (
+    ActionQueue,
+    Action,
+    Phase,
+    Player,
+    State,
+    Logic,
+    Board,
+    Cell,
+    Mark)
+from collections.abc import Sequence
+from typing import Optional
+import unittest
+
+
+class ListActionQueue(ActionQueue):
+    def __init__(self, actions: Sequence[Optional[Action]]):
+        self.actions = actions
+        self.i = 0
+
+    def next(self) -> Optional[Action]:
+        if self.i >= len(self.actions):
+            return None
+        action = self.actions[self.i]
+        self.i += 1
+        return action
+
+
+def _new_state(
+        game_rounds=5,
+        player1=None,
+        player2=None,
+        board=None,
+        phase=Phase.INROUND,
+        round_=0,
+        step=0) -> State:
+    player1 = Player(Mark.X) if player1 is None else player1
+    player2 = Player(Mark.O) if player2 is None else player2
+    board = Board() if board is None else board
+    return State(
+            game_rounds=game_rounds, players=[player1, player2], board=board, phase=phase,
+            round_=round_, step=step)
+
+
+def _new_player(mark: Mark, wins: int) -> Player:
+    player = Player(mark)
+    player.wins = wins
+    return player
+
+
+class LogicSingleActionTest(unittest.TestCase):
+    def test_advance__none_action(self):
+        state = _new_state(board=Board([
+                [None, None, None], [None, None, Mark.X], [None, None, None]]))
+        Logic([ListActionQueue([None]), ListActionQueue([None])]).advance(state)
+        expected_board_cells = [
+                [None, None, None], [None, None, Mark.X], [None, None, None]]
+        self.assertEqual(state.board.cells, expected_board_cells)
+        self.assertEqual(state.step, 0)
+        self.assertEqual(state.round, 0)
+        self.assertEqual(state.phase, Phase.INROUND)
+        self.assertEqual(state.players[0].wins, 0)
+        self.assertEqual(state.players[1].wins, 0)
+
+    def test_advance__occupy_valid_action(self):
+        state = _new_state()
+        Logic([
+                ListActionQueue([Action.new_occupy(Cell(1, 2))]),
+                ListActionQueue([None])]).advance(state)
```

Comment:
The following formatting is preferrable, provided that the linter is happy:
```python
Logic([
                ListActionQueue([Action.new_occupy(Cell(1, 2))]),
                ListActionQueue([None])])
        .advance(state)
```

Reply:
Done.
Although PyCharm linter rises weak warring: "PEP 8: E127 continuation line over-indented for visual indent" on the line with `.advance(state)`, I do not see violation PEP8 rules.